tests/unit/v1/test_bundle.py

```diff
@@ -0,0 +1,468 @@
+# Copyright 2021 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import typing
+import unittest
+
+import mock
+from google.cloud.firestore_bundle import BundleElement, FirestoreBundle
+from google.cloud.firestore_v1 import _helpers
+from google.cloud.firestore_v1.async_collection import AsyncCollectionReference
+from google.cloud.firestore_v1.base_query import BaseQuery
+from google.cloud.firestore_v1.collection import CollectionReference
+from google.cloud.firestore_v1.query import Query
+from google.cloud.firestore_v1.services.firestore.client import FirestoreClient
+from google.cloud.firestore_v1.types.document import Document
+from google.cloud.firestore_v1.types.firestore import RunQueryResponse
+from google.protobuf.timestamp_pb2 import Timestamp  # type: ignore
+from tests.unit.v1 import _test_helpers
+from tests.unit.v1 import test__helpers
+
+
+class _CollectionQueryMixin:
+
+    # Path to each document where we don't specify custom collection names or
+    # document Ids
+    doc_key: str = "projects/project-project/databases/(default)/documents/col/doc"
+
+    @staticmethod
+    def build_results_iterable(items):
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_collection_class():
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_internal_client_mock():
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_client():
+        raise NotImplementedError()
+
+    def _bundled_collection_helper(
+        self, document_ids: typing.Optional[typing.List[str]] = None
+    ) -> CollectionReference:
+        """Builder of a mocked Query for the sake of testing Bundles.
+
+        Bundling queries involves loading the actual documents for cold storage,
+        and this method arranges all of the necessary mocks so that unit tests
+        can think they are evaluating a live query.
+        """
+        client = self.get_client()
+        template = client._database_string + "/documents/col/{}"
+        document_ids = document_ids or ["doc-1", "doc-2"]
+        documents = [
+            RunQueryResponse(
+                transaction=b"",
+                document=Document(
+                    name=template.format(document_id),
+                    fields=_helpers.encode_dict({"hello": "world"}),
```

Comment:
Can we add some more data types, also include some non-ASCII chars?

Reply:
Ah yes, I tested these manually, but there should be unit tests for them as well. I will add some.